File: 239-sliding-window-maximum/sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        // intution: store a STRICTLY DECREASING deque
        // front of the deque would be the largest element in that window
        // then we will pop the front when left pointer goes past it

        deque<int> d;
        vector<int> ans;

        // populate the intial deque
        for (int i = 0; i < k; i++) {
            // don't have to keep any smaller elements than the max element in the initial window
            while (!d.empty() && nums[i] > d.back()) {
                d.pop_back();
            }
            d.push_back(nums[i]);
        }
        ans.push_back(d[0]);

        // slide the window
        int l = 0;
        for (int r = k; r < nums.size(); r++) {
            // remove front of deque if left pointer is that element
            if (nums[l] == d[0]) {
                d.pop_front();
            }
            l++;

            // add in right element
            while (!d.empty() && nums[r] > d.back()) {
                d.pop_back();
            }
            d.push_back(nums[r]);
            ans.push_back(d[0]);
        }

        return ans;
    }
};

File: 239-sliding-window-maximum/sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        # monotonic decreasing queue
        # max element should be at the front
        res = []
        q = deque()

        # initalise the monotonic decreasing queue
        for i in range(k):
            while q and nums[i] > q[-1]: # while incoming > last elemment in queue
                q.pop() # remove the last element so it will be monotonic decreasing
            q.append(nums[i])
        
        res.append(q[0])
        l,r = 0,k-1

        while r < len(nums)-1:
            # remove max element if it is no longer in window (left)
            if nums[l] == q[0]:
                q.popleft()
            l += 1
            
            # add in right element into montonic decreasing queue 
            r += 1
            while q and nums[r] > q[-1]:
                q.pop()
            q.append(nums[r])
            
            res.append(q[0])
        return res
